circuits/circuits/withdraw.circom

```diff
@@ -1,26 +1,49 @@
 pragma circom 2.1.5;
 
-include "./node_modules/circomlib/circuits/comparators.circom";
 include "./merkle_tree.circom";
+include "./sig.circom";
+include "./git_modules/circom-ecdsa/circuits/ecdsa.circom";
+include "./git_modules/circom-ecdsa/circuits/zk-identity/eth.circom";
+include "./node_modules/circomlib/circuits/comparators.circom";
 
 // If the sender is the 0 address, the transaction is a mint request,
 // if the recipient is 0, the transaction is a withdrawal, if neither are
 // 0 it's an L2->L2 send, and both can't be 0.
-template Withdraw(levels) {
+template Withdraw(levels, n, k) {
     signal input sender;
     signal input recipient;
     signal input initial_root;
 
-    signal output new_root;
-
     signal input leaf_coins[2];
     signal input pathElements[levels];
     signal input pathIndices[levels];
 
-    // make sure it's a withdraw request by checking recipient is zero and sender is non-zero.
-    recipient === 0;
-    signal is_sender_zero <== IsZero()(in <== sender);
-    is_sender_zero === 0;
+    signal output new_root;
+
+    // signature verification.
+    {
+        signal input r[k];
+        signal input s[k];
+        signal input msghash[k];
+        signal input pubkey[2][k];
+        // signature verification.
+        VerifySignature(n, k)(
+            r <== r,
+            s <== s,
+            msghash <== msghash,
+            pubkey <== pubkey,
+            leaf_coins <== leaf_coins,
+            receiver <== recipient,
+            signer <== sender
+        );
+    }
+
+    {
+        // make sure it's a withdraw request by checking recipient is zero and sender is non-zero.
+        recipient === 0;
+        signal is_sender_zero <== IsZero()(in <== sender);
+        is_sender_zero === 0;
+    }
```

Comment:
It'd be nice to allow invalid signatures here too.

Reply:
fixed.